lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -0,0 +1,148 @@
+
+        TODO: The dispatcher is currently merging JSON request body and Flask
+        URI keywords ... it probably makes sense to also merge request query
+        parameters and then validate them all together; although it might make
+        more sense to validate each against separate schemas.
+
+        GET /api/v1/datasets/list?start=1970-01-01&end=2040-12-31&owner=fred&metadata=dashboard.seen&metadata=server.deletion
+        """
+
+        # We missed automatic schema validation due to the lack of a JSON body;
+        # construct an equivalent JSON body now so we can run it through the
+        # validator.
+        json = {
+            key: request.args.getlist(key)
+            if self.schema[key].type == ParamType.LIST
+            else request.args.get(key)
+            for key in request.args.keys()
+            if key in self.schema
+        }
+
+        # Normalize and validate the keys we got via the HTTP query string.
+        # These aren't automatically validated by the superclass, so we
+        # have to do it here.
+        try:
+            new_json = self.schema.validate(json)
+        except SchemaError as e:
+            self.logger.warning("Schema validation error {}", e)
+            abort(HTTPStatus.BAD_REQUEST, message=str(e))
+
+        # Validate the authenticated user's authorization for the combination
+        # of "owner" and "access".
+        self._check_authorization(json_data.get("owner"), new_json.get("access"))
+
+        # Build a SQLAlchemy Query object expressing all of our constraints
+        #
+        # TODO: Think about what other constraints we might want to support.
+        # Including Metadata outside the Dataset table would be more difficult,
+        # but should be possible.
+        query = Database.db_session.query(Dataset)
+        if "start" in new_json and "end" in new_json:
+            self.logger.info("Adding start / end query")
+            query = query.filter(
+                Dataset.created.between(new_json["start"], new_json["end"])
+            )
+        elif "start" in new_json:
+            self.logger.info("Adding start query")
+            query = query.filter(Dataset.created >= new_json["start"])
+        elif "end" in new_json:
+            self.logger.info("Adding end query")
+            query = query.filter(Dataset.created <= new_json["end"])
+        if "name" in new_json:
+            self.logger.info("Adding name query")
+            query = query.filter(Dataset.name.contains(new_json["name"]))
+        if "controller" in new_json:
+            self.logger.info("Adding controller query")
+            query = query.filter(Dataset.controller == new_json["controller"])
+        query = self._build_sql_query(new_json, query)
+
+        # Useful for debugging, but verbose: this displays the fully expanded
+        # SQL `SELECT` statement.
+        self.logger.debug(
+            "QUERY {}", query.statement.compile(compile_kwargs={"literal_binds": True})
```

Comment:
The problem with this debug statement is that it always executes the `query.statement.compile()`.  Any way we can avoid that?